cluster-autoscaler/cloudprovider/openstack/openstack_manager_heat.go

```diff
@@ -0,0 +1,304 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package openstack
+
+import (
+	"crypto/tls"
+	"fmt"
+	"io"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/gophercloud/gophercloud"
+	"github.com/gophercloud/gophercloud/openstack"
+	"github.com/gophercloud/gophercloud/openstack/containerinfra/v1/clusters"
+	"github.com/gophercloud/gophercloud/openstack/orchestration/v1/stacks"
+	"gopkg.in/gcfg.v1"
+	netutil "k8s.io/apimachinery/pkg/util/net"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/autoscaler/cluster-autoscaler/cloudprovider"
+	"k8s.io/autoscaler/cluster-autoscaler/config"
+	certutil "k8s.io/client-go/util/cert"
+	"k8s.io/klog"
+	provider_os "k8s.io/kubernetes/pkg/cloudprovider/providers/openstack"
+)
+
+const (
+	stackStatusUpdateInProgress = "UPDATE_IN_PROGRESS"
+	stackStatusUpdateComplete   = "UPDATE_COMPLETE"
+)
+
+// StatusesPreventingUpdate is a set of statuses that would prevent
+// the cluster from successfully scaling.
+var StatusesPreventingUpdate = sets.NewString(
+	clusterStatusUpdateInProgress,
+	clusterStatusUpdateFailed,
```

Comment:
Preventing an update on update failed is already causing issues when using Magnum. We need to sort this out upstream and update here accordingly... maybe a TODO linking upstream?

Reply:
I've created the upstream story [here](https://storyboard.openstack.org/#!/story/2005056) and added a link at https://github.com/kubernetes/autoscaler/pull/1690/commits/09e464a81edfadc84288c2237f7750528d431eef#diff-d35c147e215a3b4712ad5f5cc79e6b82R49